### ISSUE TITLE
How to create a step function to use as activation function in Keras?

### ISSUE THREAD

**Issue by EmbraceLife** [keras-team/keras#7370]
I want to build a step activation function to use in keras.

Thanks to [Q&A here](https://stackoverflow.com/questions/39921607/how-to-make-a-custom-activation-function-with-only-python-in-tensorflow), I am able to build a working step function called `tf_stepy` with tensorflow. (See code below)

Now my question evolves into 
> How to make use of this `tf_stepy` activation function created in `tensorflow` to work in `keras`?

I tried to the following code to utilize `tf_stepy` in keras, but not working. Could anyone help me make it to work in keras? Thanks

    from tensorflow_step_function import tf_stepy

    def buy_hold_sell(x):
    	return tf_stepy(x)
    
    get_custom_objects().update({'custom_activation': Activation(buy_hold_sell)})

Below is the step activation function created with tensorflow

    # tensorflow_step_function.py
    import tensorflow as tf
    import keras.backend as K
    from keras.backend.tensorflow_backend import _to_tensor
    import numpy as np
    
    def stepy(x):
    	if x < 0.33:
    		return 0.0
    	elif x > 0.66:
    		return 1.0
    	else:
    		return 0.5
    
    import numpy as np
    np_stepy = np.vectorize(stepy)
    
    def d_stepy(x): # derivative
    	if x < 0.33:
    		return 0.0
    	elif x > 0.66:
    		return 1.0
    	else:
    		return 0.5
    np_d_stepy = np.vectorize(d_stepy)
    
    import tensorflow as tf
    from tensorflow.python.framework import ops
    
    np_d_stepy_32 = lambda x: np_d_stepy(x).astype(np.float32)
    
    def py_func(func, inp, Tout, stateful=True, name=None, grad=None):
    
        # Need to generate a unique name to avoid duplicates:
        rnd_name = 'PyFuncGrad' + str(np.random.randint(0, 1E+8))
    
        tf.RegisterGradient(rnd_name)(grad)  # see _MySquareGrad for grad example
        g = tf.get_default_graph()
        with g.gradient_override_map({"PyFunc": rnd_name}):
            return tf.py_func(func, inp, Tout, stateful=stateful, name=name)
    
    def tf_d_stepy(x,name=None):
        with ops.op_scope([x], name, "d_stepy") as name:
            y = tf.py_func(np_d_stepy_32,
                            [x],
                            [tf.float32],
                            name=name,
                            stateful=False)
            return y[0]
    
    def stepygrad(op, grad):
        x = op.inputs[0]
    
        n_gr = tf_d_stepy(x)
        return grad * n_gr
    
    np_stepy_32 = lambda x: np_stepy(x).astype(np.float32)
    
    def tf_stepy(x, name=None):
    
        with ops.op_scope([x], name, "stepy") as name:
            y = py_func(np_stepy_32,
                            [x],
                            [tf.float32],
                            name=name,
                            grad=stepygrad)  # <-- here's the call to the gradient
            return y[0]
    
    with tf.Session() as sess:
    
        x = tf.constant([0.2,0.7,0.4,0.6])
        y = tf_stepy(x)
        tf.initialize_all_variables().run()
    
        print(x.eval(), y.eval(), tf.gradients(y, [x])[0].eval())

**Comment by mrTsjolder** [comment]
First of all: if your function works with the backend that is used by Keras, you should be able to pass your function where an activation function is expected in Keras, e.g.

    def my_func(x):
        # my implementation ...

    model.add(Activation(my_func))

If this does not work, you might be using the wrong backend or some more information might be necessary: What kind of errors do you get? What exactly does not work? etc...
Note that something like `get_custom_objects().update({'my_func': my_func})` simply adds your function to the custom object pool, which allows you to use 

    model.add(Activation('my_func'))

It should be noted though that it is cleaner to use [custom_object_scope](https://keras.io/utils/#custom_object_scope())

    with custom_object_scope({'my_func': my_func}):
        model.add(Activation('my_func'))

Apart from that, I would like to point you to the fact that neural networks normally require differential activation functions. ReLUs show that it also works when the derivative is not well defined in some points. Your function could be considered differentiable, but the derivative would be zero everywhere. I am not sure why you use the derivative of a completely different function than your actual function. You could have your reasons of course.

PS: have another look at [stackoverflow](https://stackoverflow.com/questions/39921607/how-to-make-a-custom-activation-function-with-only-python-in-tensorflow). There are easier ways to build your own activation function with matching gradients in tensorflow, theano and keras (see the [backend](https://keras.io/backend/) module).

Hope this helps

**Comment by RyanCV** [comment]
Do we have to write the backpropagation for customized activation function in keras?

**Comment by mrTsjolder** [comment]
@RyanCV If you have written your custom activation function in the backend (theano, tensorflow or CNTK) that you want to use in Keras, Keras will use the automatic differentiation engine of that backend to compute the necessary gradients automatically.

**Comment by harshini-gadige** [comment]
Can this be closed ?
